concepts/functions/introduction.md

```diff
@@ -1,17 +1,105 @@
+Following is an example of a function with a side effect
 
-Functions are [_first class objects_][first-class objects], which means they can be assigned to variables, passed as arguments to other functions, used as a `return` value, and wrapped around/nested to form decorators and closures.
+```python
+def hello():
+    print("Hello")
 
-Python also provides many top-level [_built-in functions_][built-in functions] for common operations.
 
+hello()
+
+```
+
+The [`def`][def] keyword is used to signal the beginning of the function definition.
+`hello` is the name of the function.
+The empty parentheses represent that no values are being passed to the function.
+The body of the function is the single `print("Hello")` statement.
+`print` is also a function.
+The `"Hello"`  in the `print` function's parentheses is known as an [`argument`][arguments].
+The argument is used by the `print` function to know what to print.
+Note that the body of the function is indented.
+The indentation is important because Python relies on it to know where that block of code ends.
+The function body ends at either the end of the program or just before the next line of code that is _not_ indented.
+Since the `hello` function does not specify a `return` value, it executes its side effect - which is calling the `print` function - and returns `None`.
+Finally, we call the function by using its name and the parentheses.
+
+Following is an example of a function with a return value
+
+```python
+def hello():
+    return "Hello"
+
+
+print(hello())
+
+```
+
+The body of the function has been changed from calling `print` to returning `"Hello"`.
+The program still prints `"Hello"`, but the side effect takes place in the calling code instead of in the `hello` function.
+If the parentheses were left off from  calling the `hello` function (e.g. `print(hello)`), then `print` would output something like
+
+```
+<function hello at 0x0000026E18E93E20>
+```
+
+It's important that, even if a function accepts no arguments, the empty parentheses must be used to call it correctly.
+This is different from a language such as Ruby which does not require empty parentheses for calling a function without any arguments.
+
+A function can define zero or more [`parameters`][parameters]. A parameter defines what argument(s) the function accepts.
+
+Following is an example of a function which accepts an argument
+
+```python
+def hello(name):
+    return f"Hello, {name}"
+
+
+print(hello("Bob"))
+
+```
+
+The parameter is defined as `name`.
+`"Bob"` is the argument which the program passes to the `hello` function.
+The program outputs `Hello, Bob`.
+
+What if someone calls `hello` without passing an argument?
+The program would error with a report that an argument is missing.
+If we don't want the program  to error with no argument, we can define a [`default argument`][default arguments].
```

Comment:
```suggestion
If we don't want the program to error with no argument (_but suspect that the calling code might not supply one_), we can define a [`default argument`][default arguments].
```

Reply:
I like the direction of this suggestion, but I think better would be

If we don't want the program to error with no argument (_but want to allow the calling code to not supply one_), we can define a [`default argument`][default arguments].

Reply:
I think that the edit I made will allow you to commit the suggestion.  Let me know if it doesn't.

Reply:
Thank you for the effort, but I committed a manual edit yesterday, so this isn't needed. Sorry, I should have commented I did that!

Reply:
No worries, as long as it's done.  😄   I'll mark resolved then.